feat(datastore.snomed): Improve suggested changes in SnomedQueryOptimizer

apeteri opened the pull request.

apeteri committed:
feat(core): Add toString method to QueryExpression

apeteri committed:
feat(core): Add generic diff method to QueryExpressionDiff

It operates on the "normalized ECL" representation of queries and
preserves QueryExpression identifiers where possible.

apeteri committed:
feat(core): Extend SimpleTaxonomyGraph with a "subsumes" method

apeteri committed:
feat(snomed.datastore): Update query optimization class

apeteri committed:
test(snomed.datastore.tests): Add query optimization class test cases

apeteri committed:
refactor(snomed.datastore): Make SnomedRelationshipStats parameterized

apeteri committed:
test(snomed.datastore.tests): Add more query optimization test cases

apeteri committed:
refactor(snomed.datastore): Make SnomedHierarchyStats parameterized

apeteri committed:
fix(core): Avoid NPE in SimpleTaxonomyGraph#subsumes

apeteri committed:
fix(snomed.datastore): Move cancellation check to the proper level

apeteri committed:
fix(snomed.datastore): Fix registration of remaining exclusions

apeteri committed:
fix(snomed.datastore): Fix maximum clause count limiting expression

apeteri committed:
test(snomed.core.rest.tests): Fix SnomedOptimizationApiTest cases

apeteri committed:
fix(snomed.datastore): Add more log messages to SnomedQueryOptimizer

apeteri committed:
fix(snomed.datastore): Use explicit narrowing conversion to float

apeteri committed:
fix(snomed.datastore): Use constant to indicate empty diff result

apeteri committed:
fix(snomed.datastore): Allow optimization to progress after fit...

...threshold change; don't switch away immediately from the first
strategy

apeteri committed:
fix(snomed.datastore): Don't add relationship refinement exclusions...

...below five true positive matches

apeteri committed:
fix(snomed.datastore): Increase maximum clause count

apeteri committed:
feat(snomed.datastore): Remove root concept and Clinical finding from...

...the candidate ancestor collection in SnomedHierarchyStats, as they
impose a constant (~6 seconds) overhead on descendant count evaluation
and are unlikely to be good replacements.

apeteri committed:
feat(snomed.datastore): Skip optimization entirely if the concept set...

...to optimize contains 0 or 1 members

apeteri committed:
Merge branch '8.x' into feature/snomed-query-optimizer

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedHierarchyStats.java:
```diff
@@ -0,0 +1,360 @@
+/**
+ * @since 8.11.0
+ */
+public record SnomedHierarchyStats(
```
:woah:

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,789 @@
  */
 public final class SnomedQueryOptimizer implements QueryOptimizer {
 
+	static final int PAGE_SIZE = 10_000;
```
It would be ideal to replace this with the new constant.

apeteri committed:
fix(snomed.datastore): Ignore values in SnomedRelationshipStats

apeteri committed:
feat(snomed.datastore): Use generic concept search request to...

...speed up initial concept set evaluation in SnomedQueryOptimizer

apeteri committed:
fix(snomed.datastore): Allow inactive concepts to be returned in...

...the default implementation of ConceptSetEvaluator

apeteri committed:
fix(snomed.datastore): Select relationship value fields in...

...the default implementation of RelationshipSearchBySource

apeteri committed:
fix(snomed.datastore): Don't use lossy mode when stepping to a...

...different scoring method after staying with the same for 100
iterations

apeteri committed:
Merge remote-tracking branch 'origin/8.x' into feature/snomed-query-optimizer

apeteri committed:
feat(core): Use ID filters for "descendant" and "descendant or self"...

...expressions in ConceptSearchRequestEvaluator to speed up evaluation

apeteri committed:
fix(core): Rename parameter of helper method

apeteri committed:
Merge remote-tracking branch 'origin/8.x' into feature/snomed-query-optimizer

apeteri committed:
feat(datastore): Reuse compaction method in SnomedQueryOptimizer

apeteri committed:
fix(datastore): Keep "100%" ancestors in SnomedHierarchyStats so that...

...they can still be subjected to subsumption testing

apeteri committed:
fix(datastore): Fix small issues in SnomedQueryOptimizer

- Avoid IllegalArgumentException when only a single ancestor score
remains
- Remove the duplicate with the smaller list index
- Don't compare query expressions against themselves
- Fix test condition when looking for ancestor(-or-self) expression
pairs
- Rename local variable

apeteri committed:
fix(snomed.datastore): Exit early if no inclusions are present in...

...SnomedQueryOptimizer, similar to how value set evaluation behaves

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedHierarchyStats.java:
```diff
@@ -0,0 +1,357 @@
+		positiveDescendantsAndSelf.elementSet().remove(Concepts.ROOT_CONCEPT);
+		positiveChildren.elementSet().remove(Concepts.ROOT_CONCEPT);
+		
+		positiveDescendantsAndSelf.elementSet().remove(CLINICAL_FINDING);
```
Would it make sense to treat all top-level concepts like this?

apeteri replied:
These were the costliest to even consider given their large descendant count. `373873005 |Pharmaceutical / biologic product|` might be worth adding as another exception, but I'd leave it open for other top-level concepts.

If we want to handle this dynamically and eg. exclude all candidates above a certain amount of descendants (100 000, for example), a speed-up is available in PR https://github.com/b2ihealthcare/snow-owl/pull/1186.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedHierarchyStats.java:
```diff
@@ -0,0 +1,357 @@
+			.collect(Collectors.toList());
+	}
+
+	public LongKeyFloatMap initAncestorScores(
```
It would be great to add minimal javadoc and explanation to these public methods.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedRelationshipStats.java:
```diff
@@ -0,0 +1,165 @@
+import com.google.common.collect.Table;
+
+/**
+ * @since 8.11.0
```
Please update the since tags to 8.12.0.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
 
+	static final int PAGE_SIZE = 10_000;
+
+	public enum OptimizerStrategy {
```
Add a since tag here.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+		DEFAULT,
+		
+		SCORE_BOOST_1 {
+			// Boost scores with the fraction of the precision of direct children, for selected candidates
```
Turn these comments to enum literal javadoc, please.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+		};
+		
+		public float getFalsePositiveThreshold(final float value) {
+			// Default behavior is to disallow false positives
```
Same applies to these, but as method javadoc.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+			.setLimit(pageSize)
+			.filterByEcl(eclExpression)
+			.setFields(SnomedConceptDocument.Fields.ID)
+			.sortBy(SnomedConceptDocument.Fields.ID)
```
Just a note: the `id` field is the default sort method for all revision types.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+				.filterByCodeSystemUri(resourceUri)
+				.filterByInclusions(queryExpressions)
+				.filterByExclusions(mustNotQueryExpressions)
+				.setLimit(PAGE_SIZE)
```
It looks like only the ID is needed here, please add the `id` field selection to this request.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+	private int maxClauseCount                  = 2000;
+
+	// Run-time fields
+	private Set<String> conceptSet;
```
This is a server-side singleton instance in the current implementation (see `SnomedPlugin`). Running multiple optimizations at the same time will mess up the internals. Please refactor. Feel free to create a new instance per request.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+
+	private void filterRefinementsForInclusion(final SnomedRelationshipStats relationshipStats) {
+		// a) Less than 95% precision
+		relationshipStats.filterByPrecision(0.95f);
```
Would it make sense to move any of these magic numbers to constants? Or have them configurable?

apeteri replied:
We can expose any of the parameters as `Option` keys, and pick up defaults from eg. `SomedCoreConfiguration`, but I'm not sure who would want to adjust them. An "advanced optimize" dialog would look like a studio mixing board.

cmark replied:
It's okay to move them to named constants. I'm unsure whether having them configurable externally would be great or not.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+			relationshipSearchByTypeAndDestination);
+
+		filterRefinementsForInclusion(inclusionRelationshipStats);
+		final List<QueryExpression> refinementInclusions = inclusionRelationshipStats.optimizeRefinements(context);
```
This will always try to optimize using refinements first, right? Isn't that costly? I mean if there are 10k conceptIds then we always start with a relationship fetch according to this method call. Would it be possible to move this to a configuration value and check how long it takes to optimize a 1k and 10k refset with refinements enabled and disabled? Thanks!

apeteri replied:
Runtimes are in milliseconds:

Scenario*      | 1k disabled | 1k enabled | 10k disabled | 10k enabled
-------------- | ------------ | ------------- | -------------- | -------------
Attempt 1     | 7820           | 8097            | 59672            | 63078
Attempt 2     | 7684           | 8029            | 61261            | 61407
Attempt 3     | 7700           | 7961            | 59968            | 62501
**Average**  | 7734.67      | 8029            | 60300.33       | 62328.67

<sup>*Actual member counts were 932 (VS 999029411000230108) and 9734 (VS 999021651000230109)</sup>

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+		// See if concepts still to be processed can be converted to "* : c1 = c2" expressions
+		final SnomedRelationshipStats inclusionRelationshipStats = SnomedRelationshipStats.create(
+			context, 
+			conceptsToInclude, 
```
This line could be `conceptSet` here to be consistent with the other calls and I think by doing that the comment on line 365 is not needed anymore.

apeteri replied:
IIRC this distinction was made in the reference algorithm as well for some reason. Ancestors are allowed to cover an included concept multiple times.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+				conceptsToExclude.addAll(unexpected);
+
+				// Tune parameters after a successful selection
+				if (((float) conceptSet.size()) / (includedIds.size() + 1) > 1000.0f) {
```
What is 1000 here? 😕

apeteri replied:
If every clause would cover as many concepts from the evaluated concept set as the one that we have just found, we'd have about 1000 clauses. This is an indication that the iteration progresses too slowly.

apeteri replied:
I added comments in https://github.com/b2ihealthcare/snow-owl/commit/ce9898a94701d38b5633bef0a0b2e907a573a4bf to clarify.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+				}
+
+				// Elevate strategy after using the same one for 100 iterations (but don't promote to LOSSY from non-lossy strategies)
+				if (iteration > (optimizerStrategy.ordinal() + 1) * 100 && optimizerStrategy.ordinal() < OptimizerStrategy.values().length - 2) {
```
Ohmy, enum ordinal magic, would it be possible to refactor this to something readable? 

Something like this for example:
```
if (optimizerStrategy.isElevateNeeded(iterations)) {
  optimizerStrategy = optimizerStrategy.elevate();
}
```
And document elevation strategy in the enum type.
Thanks!

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java:
```diff
@@ -15,169 +15,847 @@
+				if (zoom <= maxClauseCount && zoom <= conceptSet.size() && zoom != Math.round(zoom * 1.1f)) {
+					// Try adjusting zoom first
+					zoom = Math.round(zoom * 1.1f);
+					log.info("Zoom changed to {}", zoom);
```
Awesome log messages, but it would be great to turn them to trace messages before we merge this PR to 8.x.

apeteri replied:
I left the initial and final messages on "info" level and turned the rest to "trace".

cmark reviewed (changes requested):
Looks great, but there are things to do.

Could you please update the document describing the summarized version of the algorithm?
Also, could you please measure the memory usage of the new algorithm for a few refsets, including larger ones with more than 10k single concept clauses? Thanks!

apeteri committed:
fix(snomed.datastore): Make page size configurable in optimizer

apeteri committed:
fix(snomed.datastore): Create a new instance for each optimization call

apeteri committed:
chore(snomed.datastore): Update "since" tags to 8.12.0

apeteri committed:
chore(snomed.datastore): Convert inline comments to javadoc for the...

...strategy enum in SnomedQueryOptimizer

apeteri committed:
refactor(snomed.datastore): Remove redundant "sortBy" call

apeteri committed:
refactor(snomed.datastore): Add field selection to generic concept...

...search request in SnomedQueryOptimizer

apeteri committed:
refactor(snomed.datastore): Use methods on OptimizerStrategy literals...

...to determine whether switching to another strategy is required

apeteri committed:
refactor(snomed.datastore): Adjust log levels in SnomedQueryOptimizer

- Clause evaluation errors are logged with "error" level, and include
the syntax exception that was thrown
- Initial and final concept/clause counts are left on "info" level
- Other internal log messages emitted during optimization are changed to
"trace" level

apeteri committed:
feat(core): Allow additional options to be passed to the optimizer

apeteri committed:
feat(snomed.datastore): Introduce option key for skipping...

...relationship refinement clause collection in SnomedQueryOptimizer

apeteri committed:
refactor(snomed.datastore): Add javadoc to SnomedRelationshipStats

apeteri committed:
refactor(snomed.datastore): Add javadoc to SnomedHierarchyStats

apeteri committed:
fix(snomed.datastore): Recompute ancestor scores each time one of its...

...input parameters change; rename "zoom" to "targetClauseCount"

apeteri commented:
With the 1k value set, heap usage is around 300 MB:

![image](https://github.com/b2ihealthcare/snow-owl/assets/1411829/db4df2da-1696-45db-9aea-5339115f737c)

I could not generate a chart for the 10k version, because with instrumentation the optimizing process takes a lot longer to run. But it seems that one goes up to 1.7 GB of heap usage.

apeteri committed:
fix(snomed.datastore): Don't remove pinned expression when compacting

apeteri committed:
fix(core): Rewrite the value portion of ECL attribute comparisons

cmark reviewed (approved):
🥓

cmark commented:
> With the 1k value set, heap usage is around 300 MB:
> 
> ![image](https://user-images.githubusercontent.com/1411829/256494228-db4df2da-1696-45db-9aea-5339115f737c.png)
> 
> I could not generate a chart for the 10k version, because with instrumentation the optimizing process takes a lot longer to run. But it seems that one goes up to 1.7 GB of heap usage.

Hmm, that's a lot 😄 Would it be possible to identify areas where memory consumption could be lowered a bit? No need to refactor anything now, just mark variables, statements with a comment where optimization could take place.

apeteri commented:
I suspect most of that is going in the ECL evaluation cache, as the process invokes the evaluator with different expressions in quick succession. I will need to fire up a profiler to dig into allocation details.

cmark commented:
> I suspect most of that is going in the ECL evaluation cache, as the process invokes the evaluator with different expressions in quick succession. I will need to fire up a profiler to dig into allocation details.

Did you measure this with a dedicated Elasticsearch instance or in embedded mode?

apeteri commented:
ES was running separately, I'm thinking of `SnomedConceptRequestCache` here... maybe?

cmark commented:
> `SnomedConceptRequestCache`

That's only for data expansion queries, ECL caches are outside of Snow Owl.
